lib/crypto/c_src/api_ng.c

```diff
@@ -0,0 +1,155 @@
+        return ERROR_Str(env, "Unknown cipher or bad key size");
+
+    if (FORBIDDEN_IN_FIPS(cipherp))
+        return enif_raise_exception(env, atom_notsup);
+
+    if (!(cipher = cipherp->cipher.p)) {
+#if !defined(HAVE_EVP_AES_CTR)
+        if (cipherp->flags & AES_CTR_COMPAT)
+            return aes_ctr_stream_init_compat(env, argv[1], argv[2]);
+        else
+#endif
+            return enif_raise_exception(env, atom_notsup);
+    }
+
+    iv_len = EVP_CIPHER_iv_length(cipher);
+
+    if (iv_len) {
+        if (   !enif_inspect_binary(env, argv[2], &ivec_bin)
+            || (iv_len != ivec_bin.size)
+           ) {
+            return ERROR_Str(env, "Bad iv size");
+        }
+        iv = ivec_bin.data;
+    }
+
+    ctx = enif_alloc_resource(evp_cipher_ctx_rtype, sizeof(struct evp_cipher_ctx));
+    ctx->ctx = EVP_CIPHER_CTX_new();
+
+    if (!EVP_CipherInit_ex(ctx->ctx, cipher, NULL, key_bin.data, iv, enc))
+        return ERROR_Str(env, "Can't initialize key and/or iv");
+
+    EVP_CIPHER_CTX_set_padding(ctx->ctx, 0);
+
+    ret = enif_make_resource(env, ctx);
+    enif_release_resource(ctx);
+    return ret;
+}
+
+ERL_NIF_TERM ng_crypto_flag_nif(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
+{/* (Context, EncodeFlg) */
+    struct evp_cipher_ctx *ctx;
+    int ok = 1;
+
+    if (!enif_get_resource(env, argv[0], evp_cipher_ctx_rtype, (void**)&ctx))
+        return enif_make_badarg(env);
+
+    if (argv[2] != atom_undefined) {
+        /* Couldn't set the encrypt flag in the init call (= compatibility routines), so do
+           it now
+        */
+        if (argv[2] == atom_true)
+            ok = EVP_CipherInit_ex(ctx->ctx, EVP_CIPHER_CTX_cipher(ctx->ctx), NULL, NULL, NULL, 1);
+        else if (argv[2] == atom_false)
+            ok = EVP_CipherInit_ex(ctx->ctx, EVP_CIPHER_CTX_cipher(ctx->ctx), NULL, NULL, NULL, 0);
+        else
+            return ERROR_Term(env, enif_make_atom(env,"badarg"));
+ }
+
+ if (!ok)
+     return ERROR_Str(env, "Couldn't set encrypt/decrypt flag");
+
+ return atom_ok;
+}
+
+
+ERL_NIF_TERM ng_crypto_update_nif(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
+{/* (Context, Data) */
+    struct evp_cipher_ctx *ctx;
+    ErlNifBinary   data_bin;
+    ERL_NIF_TERM   cipher_term;
+    unsigned char  *out;
+    int            outl = 0;
+
+#if !defined(HAVE_EVP_AES_CTR)
+    const ERL_NIF_TERM *state_term;
+    int state_arity;
+
+    if (enif_get_tuple(env, argv[0], &state_arity, &state_term) && (state_arity == 4)) {
+        return aes_ctr_stream_encrypt_compat(env, argv[0], argv[1]);
+    }
+#endif
+
+    if (!enif_get_resource(env, argv[0], evp_cipher_ctx_rtype, (void**)&ctx)
+        || !enif_inspect_iolist_as_binary(env, argv[1], &data_bin)) {
+        return ERROR_Term(env, enif_make_atom(env,"badarg"));
+    }
+
+    out = enif_make_new_binary(env, data_bin.size, &cipher_term);
+    if (!EVP_CipherUpdate(ctx->ctx, out, &outl, data_bin.data, data_bin.size))
```

Comment:
This is going to silently truncate because EVP_CipherUpdate() expects an int size.